### DISCUSSION TITLE
Avalonia UI Custom Elements / Web Components

### DISCUSSION THREAD

**Discussion by zborowa** [AvaloniaUI/Avalonia#12755] [Q&A] [answered]
Is it possible to create web components out of Avalonia user controls as you can do with Blazor?

```
// Example of registering a Blazor component as a web component for other applications to make use off
builder.RootComponents.RegisterCustomElement<MyCustomElement>("my-custom-element");
```

I was wondering if it would be possible to embed an Avalonia page or element inside of other web applications like Angular or React for transitional purposes. I already have a large Angular application that I would like to slowly transition to Avalonia page by page. With desktop application you can potentially just statically bind to compiled DLLs or even add the avalonia libraries to your application and that way import/use the user controls but how about web apps?

**Comment by maxkatz6** [comment, marked as answer]
There is no API to register custom components like this, but you can achieve it in a different way.

For Avalonia to run all you need is a named `div`, and pass its name to the Avalonia initialization code.

In default templates (as well as ControlCatalog app), we have a div named "out" - https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog.Browser/index.html#L17
And we pass this name to the initialization code: https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog.Browser/Program.cs#L20

Alternatively, you can set up browser app without passing any div, and instead create one or multiple AvaloniaViews later, like this in this commented code: https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog.Browser/Program.cs#L23-L32

**Reply by zborowa** [reply]
I see, that's interesting how it could potentially work. What this means is that the AppBundle could just be served with the wrapping app and that you could include it in one of the Angular components, by it I mean the javascript files (main, dotnet, avalonia).
I've managed to get the Ava app rendered in an Angular startup app and it looks pretty good, I would say that the rendered result is a bit blurry compared to running the app standalone (not wrapped) and it's clear that the canvas expects to be in the top left corner of the screen because of a mouse click offset problem.
In the following example I took the standard Angular app template and added the div out before the angular app root element in the index.html and because of the standard angular banner I added some spacing between the top and the div out which shows the mouse click offset problem. Any ideas on the two problems that I've mentioned?

![ava-canvas-button-in-angular-app](https://github.com/AvaloniaUI/Avalonia/assets/2786784/27756693-ad65-40ab-a1b6-84528cafb74b)

If it would be useful I could create an example repo with this code for reference.

**Reply by maxkatz6** [reply]
Not sure about blur, but mouse handling is definitely a bug. Can you open an issue for that?

**Reply by zborowa** [reply]
Added an issue for the mouse handling. https://github.com/AvaloniaUI/Avalonia/issues/12780
I'll have to test a bit more regarding the blur to make it clear why and when it happens. I think it's mainly when the div out is not full page wide so any embedded application will be slightly blurry because of wonky interpolation.

**Comment by kekekeks** [comment]
Note that you need to reuse the Avalonia-driven element, since there is a limit of 8 WebGL contexts per page.

**Reply by zborowa** [reply]
What do you mean exactly by reuse? Like in the example of maxkatz6 where you can create multiple AvaloniaViews and somehow render a different one in the same element depending on the page you're viewing?

**Reply by maxkatz6** [reply]
Yeah, browser doesn't allow more than 8 opengl canvases at a time.
